clippy_lints/src/implicit_saturating_sub.rs

```diff
@@ -0,0 +1,173 @@
+
+            // Check if assign operation is done
+            if let StmtKind::Semi(ref e) = block.stmts[0].kind;
+            if let (true, Some(target)) = subtracts_one(cx, e);
+
+            // Extracting out the variable name
+            if let ExprKind::Path(ref assign_path) = target.kind;
+            if let QPath::Resolved(_, ref ares_path) = assign_path;
+
+            then {
+                // Handle symmetric conditions in the if statement
+                let (cond_var, cond_num_val) = if SpanlessEq::new(cx).eq_expr(cond_left, target) {
+                    if BinOpKind::Gt == cond_op.node || BinOpKind::Ne == cond_op.node {
+                        (cond_left, cond_right)
+                    } else {
+                        return;
+                    }
+                } else if SpanlessEq::new(cx).eq_expr(cond_right, target) {
+                    if BinOpKind::Lt == cond_op.node || BinOpKind::Ne == cond_op.node {
+                        (cond_right, cond_left)
+                    } else {
+                        return;
+                    }
+                } else {
+                    return;
+                };
+
+                // Check if the variable in the condition statement is an integer
+                if !cx.tables.expr_ty(cond_var).is_integral() {
+                    return;
+                }
+
+                // Get the variable name
+                let var_name = ares_path.segments[0].ident.name.as_str();
+                const INT_TYPES: [&str; 5] = ["i8", "i16", "i32", "i64", "i128"];
+
+                match cond_num_val.kind {
+                    ExprKind::Lit(ref cond_lit) => {
+                        // Check if the constant is zero
+                        if let LitKind::Int(0, _) = cond_lit.node {
+                            if cx.tables.expr_ty(cond_left).is_signed() {
+                            } else {
+                                print_lint_and_sugg(cx, &var_name, expr);
+                            };
+                        }
+                    },
+                    ExprKind::Path(ref cond_num_path) => {
+                        if INT_TYPES.iter().any( |int_type| match_qpath(cond_num_path, &[int_type, "MIN"])) {
+                            print_lint_and_sugg(cx, &var_name, expr);
+                        };
+                    },
+                    ExprKind::Call(ref func, _) => {
+                        if let ExprKind::Path(ref cond_num_path) = func.kind {
+                            if INT_TYPES.iter().any( |int_type| match_qpath(cond_num_path, &[int_type, "min_value"])) {
+                                print_lint_and_sugg(cx, &var_name, expr);
+                            }
+                        };
+                    },
+                    _ => (),
+                }
+            }
+        }
+    }
+}
+
+fn subtracts_one<'a>(cx: &LateContext<'_, '_>, expr: &Expr<'a>) -> (bool, Option<&'a Expr<'a>>) {
```

Comment:
You always return `(true, Some)` or `(false, None)`, so returning an option is enough here.

(That would be a great lint 😄)

Reply:
Yes :laughing:
Wow umm I didn't think of it that way!